### DISCUSSION TITLE
How to start Vitest UI through Node API

### DISCUSSION THREAD

**Discussion by guoyunhe** [vitest-dev/vitest#5064] [Q&A] [answered]
I am making a CLI tool which uses Jest and Vitest Node API (user can choose what they prefer) to perform test command. Here is my code:

```ts
import jest from 'jest';
import { createVitest } from 'vitest/node';
import { loadConfig } from '../config.js';

export interface TestOptions {
  watch?: boolean;
  ui?: boolean;
}

export async function testAction({ watch, ui }: TestOptions) {
  const config = await loadConfig();
  if (config.test?.framework === 'vitest') {
    const vitest = await createVitest('test', {
      watch,
      ui,
      passWithNoTests: true,
      globals: true,
      environment: 'happy-dom',
      coverage: {
        enabled: true,
        provider: 'istanbul',
        reporter: ['html', 'json', 'json-summary', 'text'],
      },
    });

    await vitest.start();
    if (!watch && !ui) {
      await vitest.close();
    }
  } else {
    jest.run(process.argv.slice(3));
  }
}
```

When I run `my-command test --ui`, the web ui cannot connect to server:

![image](https://github.com/vitest-dev/vitest/assets/5836790/3a0faf58-b5b9-46b7-b2b1-fb90cc413fbd)

**Comment by hi-ogawa** [comment]
I tested simplified version of your code and Vitest UI looks working fine to me (tested both on stackblitz and locally):

https://stackblitz.com/edit/vitest-dev-vitest-9prb1m?file=repro.js

<details><summary>Show screenshot</summary>

![image](https://github.com/vitest-dev/vitest/assets/4232207/6e8511aa-444b-41f6-a570-e9d61dd65bcf)

</details>

I don't know but maybe your vite config might be affecting server setting.
Does UI work when you run tests from standard `vitest` cli?

**Reply by guoyunhe** [reply, marked as answer]
@hi-ogawa thank your for the help! i found the issue. In my case, I read --ui and --watch option form command line manually. when I run `my-command test --ui`, it actually gives:

```js
const vitest = await createVitest('test', {
  watch: false,
  ui: true
});
```

which is an invalid option combination. (not sure if this is considered as a bug?)

I changed my code a little bit, now it works:

```ts
export async function testAction({ watch, ui }: TestOptions) {
  const config = await loadConfig();
  if (config.test?.framework === 'vitest') {
    const vitest = await createVitest('test', {
      watch: watch || ui, // 👈 this is the change
      ui,
      passWithNoTests: true,
      globals: true,
      environment: 'happy-dom',
      coverage: {
        enabled: true,
        provider: 'istanbul',
        reporter: ['html', 'json', 'json-summary', 'text'],
      },
    });

    await vitest.start();
    if (!watch && !ui) {
      await vitest.close();
    }
  } else {
    jest.run(process.argv.slice(3));
  }
}
```
